### PR TITLE
Feature/survive plot api

### PR DESCRIPTION
**Pull request by andricDu**
API now excludes donors present in the intersection of the sets. Also filters for donors that have survival data and status.

### CODE REVIEW

File: dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/SurvivalAnalyzer.java [hunk 1]
```diff
@@ -40,34 +47,37 @@
...
-          .size(20000)
-          .sort("survivalTime")
-          .order("asc")
-          .build();
-
-      val result = donorRepository.findAllCentric(query);
-      val intervals = compute(result.getHits().getHits());
-
-      analysis.getResults().add(analysis.new Result(id, intervals));
+    val sets = analysis.getEntitySetIds();
```
Comment (btiernay) [inline dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/SurvivalAnalyzer.java]: I would make this for loop a method.

File: dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/SurvivalAnalyzer.java [hunk 2]
```diff
@@ -40,34 +47,37 @@
...
+      val exclusions = sets.stream()
+          .filter(s -> !s.equals(id))
+          .collect(toImmutableSet());
+
+      entitySetMapBuilder.put(id, new UnionUnit(ImmutableSet.<UUID> of(id), exclusions));
+    }
+
+    val entitySetMap = entitySetMapBuilder.build();
+
+    for (val e : entitySetMap.entrySet()) {
```
Comment (btiernay) [inline dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/SurvivalAnalyzer.java]: Same

File: dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/SurvivalAnalyzer.java [hunk 3]
```diff
@@ -142,6 +152,11 @@ public SurvivalAnalysis analyze(SurvivalAnalysis analysis) {
     return intervals;
   }
 
+  private static boolean hasData(SearchHit donor) {
+    return ((String) donor.field("donor_vital_status").getValue()).equalsIgnoreCase("alive") ||
```
Comment (btiernay) [inline dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/SurvivalAnalyzer.java]: Capture repeated expression as variable

File: dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/UnionAnalyzer.java [hunk 4]
```diff
@@ -220,4 +226,15 @@ private SearchResponse unionAll(final Iterable<UnionUnit> definitions, final Bas
     return response;
   }
 
+  private SearchResponse unionOne(final UnionUnit definition, final BaseEntitySet.Type entityType,
+      final int max) {
+    val response = termsLookupRepository.runUnionEsQueryWithFields(
```
Comment (btiernay) [inline dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/UnionAnalyzer.java]: Hmm not sure this method really belongs on this class. What does it do?

Comment (btiernay) [inline dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/UnionAnalyzer.java]: In terms of cohesion, sets should probably be kept out of terms lookup

File: dcc-portal-api/src/main/java/org/icgc/dcc/portal/repository/TermsLookupRepository.java [hunk 5]
```diff
@@ -209,6 +207,21 @@ public SearchResponse runUnionEsQuery(final String indexTypeName, @NonNull final
         .setNoFields());
   }
 
+  /**
+   * Special case for Survival Analysis, the fields selected for return are the only ones we currently care about.
+   */
+  public SearchResponse runUnionEsQueryWithFields(final String indexTypeName, @NonNull final SearchType searchType,
+      @NonNull final BoolFilterBuilder boolFilter, final int max) {
+    val query = QueryBuilders.filteredQuery(MATCH_ALL, boolFilter);
```
Comment (btiernay) [inline dcc-portal-api/src/main/java/org/icgc/dcc/portal/repository/TermsLookupRepository.java]: Static import

File: dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/UnionAnalyzer.java [hunk 6]
```diff
@@ -137,6 +137,12 @@ public void calculateUnionUnitCounts(@NonNull final UUID id, @NonNull final Unio
     return SearchResponses.getHitIds(response);
   }
 
+  public SearchResponse computeSetOperation(@NonNull final UnionUnit unionUnit, BaseEntitySet.Type type) {
```
Comment (btiernay) [inline dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/UnionAnalyzer.java]: `calculate/computeUnion`?

File: dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/UnionAnalyzer.java (hunk 4 above)
Comment (andricDu) [inline dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/UnionAnalyzer.java]: Would you rather put this in DonorRepository? It seems all the set related data access is stored in the TermsLookupRepository as the terms lookup are not limited to a type.

Comment (btiernay) [inline dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/UnionAnalyzer.java]: I just took a quick look there. Correct me if I'm wrong, but it looks like more than half of that class has methods which have nothing to do with terms lookup. This should just encapsulate terms lookup operations.

Comment (andricDu) [inline dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/UnionAnalyzer.java]: You are right, this probably needs a refactor then. With that said, all the public methods not responsible for create and read of a TermsLookup are passed query filters which contain terms lookup filters.

Comment (andricDu) [inline dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/UnionAnalyzer.java]: Maybe we need a `SetService` ?

Comment (btiernay) [inline dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/UnionAnalyzer.java]: You mean `EntitySetService`? Is there a difference? And ya, looks like it. Also, some of these methods should probably go into utility classes or various other services. Some filing is required.

Comment (andricDu) [inline dcc-portal-api/src/main/java/org/icgc/dcc/portal/analysis/UnionAnalyzer.java]: derp, you are right, that is a reasonable place.

**Pull request merged**